flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/BatchTableProject.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.table.plan.nodes.dataset
+
+import java.util
+
+import org.apache.calcite.plan.{RelOptCluster, RelOptCost, RelOptPlanner, RelTraitSet}
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.rel.`type`.RelDataType
+import org.apache.calcite.rel.core.Project
+import org.apache.calcite.rel.metadata.RelMetadataQuery
+import org.apache.calcite.rex.{RexInputRef, RexNode}
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.java.DataSet
+import org.apache.flink.api.table.BatchTableEnvironment
+import org.apache.flink.api.table.plan.schema.TableSourceTable
+import org.apache.flink.api.table.sources.ProjectableTableSource
+
+import scala.collection.JavaConverters._
+
+class BatchTableProject(
+    cluster: RelOptCluster,
+    traits: RelTraitSet,
+    input: RelNode,
+    projects: util.List[_ <: RexNode],
+    projectionRowType: RelDataType)
+  extends Project(cluster, traits, input, projects, projectionRowType)
+  with DataSetRel {
+
+  override def copy(traitSet: RelTraitSet, input: RelNode,
+    projects: util.List[RexNode], scanRawType: RelDataType): Project = {
+    new BatchTableProject(cluster, traitSet, input, projects, projectionRowType)
+  }
+
+  override def computeSelfCost(planner: RelOptPlanner, mq: RelMetadataQuery): RelOptCost = {
+    super.computeSelfCost(planner, mq).multiplyBy(0.8)
+  }
+
+  override def translateToPlan(
+    tableEnv: BatchTableEnvironment,
+    expectedType: Option[TypeInformation[Any]]): DataSet[Any] = {
+
+    val projectableSource = getInput.getTable.unwrap(classOf[TableSourceTable])
+      .tableSource.asInstanceOf[ProjectableTableSource[_]]
+
+    val indexes = projects.asScala.map(_.asInstanceOf[RexInputRef].getIndex).toArray
+    projectableSource.setProjection(indexes)
+
+    getInput.asInstanceOf[DataSetRel].translateToPlan(tableEnv)
```

Comment:
@wuchong, actually, `projection` works right after `tableSourceScan`. it just pushes `fieldMask` into `source` and prunes columns for logic plan. So we don't need to add any type convertions